File: sdks/typescript/src/exec.ts
import { ShellImpl, type Shell, type ShellOpts } from "./shell.js";

export interface ProcessResult {
  exitCode: number;
  stdout: string;
  stderr: string;
}

export interface RunOpts {
  timeout?: number;
  env?: Record<string, string>;
  cwd?: string;
}

export interface ExecStartOpts {
  args?: string[];
  env?: Record<string, string>;
  cwd?: string;
  timeout?: number;
  maxRunAfterDisconnect?: number;
  onStdout?: (data: Uint8Array) => void;
  onStderr?: (data: Uint8Array) => void;
  onExit?: (exitCode: number) => void;
  onScrollbackEnd?: () => void;
}

export interface ExecAttachOpts {
  onStdout?: (data: Uint8Array) => void;
  onStderr?: (data: Uint8Array) => void;
  onExit?: (exitCode: number) => void;
  onScrollbackEnd?: () => void;
}

export interface ExecSessionInfo {
  sessionID: string;
  sandboxID: string;
  command: string;
  args: string[];
  running: boolean;
  exitCode?: number;
  startedAt: string;
  attachedClients: number;
}

export interface ExecSession {
  sessionId: string;
  /** Resolves with the exit code when the process exits. */
  done: Promise<number>;
  sendStdin(data: string | Uint8Array): void;
  kill(signal?: number): Promise<void>;
  close(): void;
}

export class Exec {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private sandboxId: string,
    private token: string = "",
  ) {}

  private get headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.token) {
      h["Authorization"] = `Bearer ${this.token}`;
    } else if (this.apiKey) {
      h["X-API-Key"] = this.apiKey;
    }
    return h;
  }

  async start(command: string, opts: ExecStartOpts = {}): Promise<ExecSession> {
    const body: Record<string, unknown> = { cmd: command };
    if (opts.args) body.args = opts.args;
    if (opts.env) body.envs = opts.env;
    if (opts.cwd) body.cwd = opts.cwd;
    if (opts.timeout != null) body.timeout = opts.timeout;
    if (opts.maxRunAfterDisconnect != null) body.maxRunAfterDisconnect = opts.maxRunAfterDisconnect;

    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/exec`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to create exec session: ${resp.status} ${text}`);
    }

    const data = await resp.json();
    const sessionId: string = data.sessionID;

    return this.attach(sessionId, {
      onStdout: opts.onStdout,
      onStderr: opts.onStderr,
      onExit: opts.onExit,
      onScrollbackEnd: opts.onScrollbackEnd,
    });
  }

  async attach(sessionId: string, opts: ExecAttachOpts = {}): Promise<ExecSession> {
    const wsUrl = this.apiUrl
      .replace("http://", "ws://")
      .replace("https://", "wss://");
    // WebSocket can't set custom headers, so pass credentials as query params.
    // Prefer JWT token (direct worker access); fall back to API key (control plane).
    const authParam = this.token
      ? `?token=${encodeURIComponent(this.token)}`
      : this.apiKey
        ? `?api_key=${encodeURIComponent(this.apiKey)}`
        : "";
    const wsEndpoint = `${wsUrl}/sandboxes/${this.sandboxId}/exec/${sessionId}${authParam}`;

    const ws = new WebSocket(wsEndpoint);
    ws.binaryType = "arraybuffer";

    let gotExit = false;
    let opened = false;
    let resolveDone: (code: number) => void;
    const done = new Promise<number>((resolve) => { resolveDone = resolve; });

    await new Promise<void>((resolve, reject) => {
      ws.onopen = () => {
        opened = true;
        resolve();
      };
      ws.onerror = () => {
        if (!opened) reject(new Error(`WebSocket connection failed: ${wsEndpoint}`));
      };
      ws.onclose = () => {
        if (!opened) reject(new Error(`WebSocket closed before opening: ${wsEndpoint}`));
      };
    });

    ws.onmessage = (event) => {
      const buf = new Uint8Array(event.data as ArrayBuffer);
      if (buf.length < 1) return;

      const streamId = buf[0];
      const payload = buf.slice(1);

      switch (streamId) {
        case 0x01: // stdout
          opts.onStdout?.(payload);
          break;
        case 0x02: // stderr
          opts.onStderr?.(payload);
          break;
        case 0x03: // exit
          gotExit = true;
          if (payload.length >= 4) {
            const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
            const exitCode = view.getInt32(0, false); // big-endian
            opts.onExit?.(exitCode);
            resolveDone(exitCode);
          } else {
            opts.onExit?.(0);
            resolveDone(0);
          }
          break;
        case 0x04: // scrollback_end
          opts.onScrollbackEnd?.();
          break;
      }
    };

    ws.onclose = () => {
      if (!gotExit) {
        opts.onExit?.(-1);
        resolveDone(-1);
      }
    };

    ws.onerror = () => {
      // Post-open errors are followed by onclose, which handles exit.
    };

    const exec = this;

    return {
      sessionId,
      done,
      sendStdin(data: string | Uint8Array): void {
        if (ws.readyState !== WebSocket.OPEN) return;
        const payload = typeof data === "string"
          ? new TextEncoder().encode(data)
          : data;
        const msg = new Uint8Array(1 + payload.length);
        msg[0] = 0x00; // stdin stream ID
        msg.set(payload, 1);
        ws.send(msg);
      },
      async kill(signal?: number): Promise<void> {
        await exec.kill(sessionId, signal);
      },
      close(): void {
        ws.close();
      },
    };
  }

  async list(): Promise<ExecSessionInfo[]> {
    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/exec`, {
      headers: this.headers,
    });

    if (!resp.ok) {
      throw new Error(`Failed to list exec sessions: ${resp.status}`);
    }

    return resp.json();
  }

  async kill(sessionId: string, signal?: number): Promise<void> {
    const body: Record<string, unknown> = {};
    if (signal != null) body.signal = signal;

    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/exec/${sessionId}/kill`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to kill exec session: ${resp.status} ${text}`);
    }
  }

  /**
   * Alias for {@link start}. Use when the intent is "run this command in the
   * background and observe it" rather than the more ambiguous "start". Same
   * options, same return type.
   */
  async background(command: string, opts: ExecStartOpts = {}): Promise<ExecSession> {
    return this.start(command, opts);
  }

  /**
   * Open a stateful shell session. Subsequent `.run()` calls share the same
   * bash process, so `cwd`, exported env vars, and shell functions persist
   * across calls — the ergonomics of a terminal tab.
   *
   * Backed by a long-running exec session running `bash --noprofile --norc`.
   * Foreground-only: concurrent `.run()` rejects. Use `exec.background()` for
   * fire-and-forget processes. If the user command calls `exit`, the shell
   * closes (same as closing a terminal tab) and subsequent `.run()` rejects.
   */
  async shell(opts: ShellOpts = {}): Promise<Shell> {
    let impl: ShellImpl | null = null;
    const session = await this.start("bash", {
      args: ["--noprofile", "--norc", "+m"],
      env: opts.env,
      cwd: opts.cwd,
      onStdout: (chunk) => impl?.onStdoutChunk(chunk),
      onStderr: (chunk) => impl?.onStderrChunk(chunk),
      onScrollbackEnd: () => impl?.onScrollbackEnd(),
    });
    impl = new ShellImpl(session);
    return impl;
  }

  /**
   * Re-attach to a shell session that was previously opened by `shell()` and
   * whose sessionId you've kept. Useful for revisiting a long-lived terminal
   * tab from a different process invocation.
   *
   * Assumes the shell is idle (no in-flight `.run()` from another client).
   * If another client has a run in flight, output will interleave and the
   * results are undefined — coordinate at the application level.
   */
  async reattachShell(sessionId: string): Promise<Shell> {
    let impl: ShellImpl | null = null;
    const session = await this.attach(sessionId, {
      onStdout: (chunk) => impl?.onStdoutChunk(chunk),
      onStderr: (chunk) => impl?.onStderrChunk(chunk),
      onScrollbackEnd: () => impl?.onScrollbackEnd(),
    });
    impl = new ShellImpl(session);
    return impl;
  }

  async run(command: string, opts: RunOpts = {}): Promise<ProcessResult> {
    const body: Record<string, unknown> = {
      cmd: "sh",
      args: ["-c", command],
    };
    if (opts.env) body.envs = opts.env;
    if (opts.cwd) body.cwd = opts.cwd;
    if (opts.timeout != null) body.timeout = opts.timeout;
    else body.timeout = 60;

    const resp = await fetch(`${this.apiUrl}/sandboxes/${this.sandboxId}/exec/run`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to run command: ${resp.status} ${text}`);
    }

    return resp.json();
  }
}

File: sdks/rust/src/exec.rs
use std::collections::HashMap;
use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use tokio::sync::{mpsc, oneshot, Mutex};
use tokio::task::JoinHandle;
use tokio_tungstenite::tungstenite::Message;

use crate::error::Result;
use crate::sandbox::{check_ok, parse_response, ClientCtx};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessResult {
    #[serde(rename = "exitCode")]
    pub exit_code: i32,
    #[serde(default)]
    pub stdout: String,
    #[serde(default)]
    pub stderr: String,
}

#[derive(Debug, Clone, Default)]
pub struct RunOpts {
    pub timeout: Option<u64>,
    pub env: Option<HashMap<String, String>>,
    pub cwd: Option<String>,
}

impl RunOpts {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn timeout(mut self, secs: u64) -> Self {
        self.timeout = Some(secs);
        self
    }

    pub fn env(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.env
            .get_or_insert_with(HashMap::new)
            .insert(key.into(), value.into());
        self
    }

    pub fn envs(mut self, env: HashMap<String, String>) -> Self {
        self.env = Some(env);
        self
    }

    pub fn cwd(mut self, cwd: impl Into<String>) -> Self {
        self.cwd = Some(cwd.into());
        self
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecSessionInfo {
    #[serde(rename = "sessionID")]
    pub session_id: String,
    #[serde(rename = "sandboxID", default)]
    pub sandbox_id: String,
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    pub running: bool,
    #[serde(rename = "exitCode", default)]
    pub exit_code: Option<i32>,
    #[serde(rename = "startedAt", default)]
    pub started_at: String,
    #[serde(rename = "attachedClients", default)]
    pub attached_clients: u32,
}

#[derive(Debug, Clone, Default)]
pub struct ExecStartOpts {
    pub args: Option<Vec<String>>,
    pub env: Option<HashMap<String, String>>,
    pub cwd: Option<String>,
    pub timeout: Option<u64>,
    pub max_run_after_disconnect: Option<u64>,
}

impl ExecStartOpts {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn args(mut self, args: Vec<String>) -> Self {
        self.args = Some(args);
        self
    }

    pub fn env(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.env
            .get_or_insert_with(HashMap::new)
            .insert(key.into(), value.into());
        self
    }

    pub fn cwd(mut self, cwd: impl Into<String>) -> Self {
        self.cwd = Some(cwd.into());
        self
    }

    pub fn timeout(mut self, secs: u64) -> Self {
        self.timeout = Some(secs);
        self
    }
}

/// Channel of streaming events from an attached exec session.
pub enum StreamEvent {
    Stdout(Vec<u8>),
    Stderr(Vec<u8>),
    /// Server has finished replaying scrollback; subsequent stdout/stderr is live.
    ScrollbackEnd,
    /// Process exited with this code. Always the last event.
    Exit(i32),
}

pub struct ExecSession {
    pub session_id: String,
    pub sandbox_id: String,
    exit: Mutex<Option<oneshot::Receiver<i32>>>,
    cached_exit: Mutex<Option<i32>>,
    stdin_tx: mpsc::UnboundedSender<Vec<u8>>,
    reader: Mutex<Option<JoinHandle<()>>>,
    ctx: Arc<ClientCtx>,
}

impl ExecSession {
    /// Wait for the process to exit and return its exit code.
    pub async fn done(&self) -> i32 {
        if let Some(code) = *self.cached_exit.lock().await {
            return code;
        }
        let recv = self.exit.lock().await.take();
        let code = match recv {
            Some(rx) => rx.await.unwrap_or(-1),
            None => -1,
        };
        *self.cached_exit.lock().await = Some(code);
        code
    }

    /// Write to the process stdin.
    pub fn send_stdin(&self, data: impl Into<Vec<u8>>) {
        let _ = self.stdin_tx.send(data.into());
    }

    /// Kill the underlying process. Default signal is SIGKILL (9).
    pub async fn kill(&self, signal: Option<i32>) -> Result<()> {
        let resp = self
            .ctx
            .http
            .post(format!(
                "{}/sandboxes/{}/exec/{}/kill",
                self.ctx.api_url, self.sandbox_id, self.session_id
            ))
            .headers(self.ctx.headers())
            .json(&json!({ "signal": signal.unwrap_or(9) }))
            .send()
            .await?;
        check_ok(resp, "kill exec session").await
    }

    /// Detach from the session. Does not kill the process.
    pub async fn close(&self) {
        if let Some(handle) = self.reader.lock().await.take() {
            handle.abort();
        }
    }
}

pub struct Exec {
    ctx: Arc<ClientCtx>,
    sandbox_id: String,
}

impl Exec {
    pub(crate) fn new(ctx: Arc<ClientCtx>, sandbox_id: String) -> Self {
        Self { ctx, sandbox_id }
    }

    /// Run a shell command and wait for completion.
    ///
    /// The command is executed via `sh -c`, so shell features like pipes,
    /// redirects, and env var expansion work as expected.
    pub async fn run(&self, command: &str, opts: RunOpts) -> Result<ProcessResult> {
        let mut body = serde_json::Map::new();
        body.insert("cmd".into(), Value::String("sh".into()));
        body.insert("args".into(), json!(["-c", command]));
        body.insert("timeout".into(), json!(opts.timeout.unwrap_or(60)));
        if let Some(env) = opts.env {
            body.insert("envs".into(), serde_json::to_value(env)?);
        }
        if let Some(cwd) = opts.cwd {
            body.insert("cwd".into(), Value::String(cwd));
        }

        let resp = self
            .ctx
            .http
            .post(format!(
                "{}/sandboxes/{}/exec/run",
                self.ctx.api_url, self.sandbox_id
            ))
            .headers(self.ctx.headers())
            .json(&Value::Object(body))
            .send()
            .await?;

        parse_response(resp, "run command").await
    }

    /// Start a long-running command and attach for streaming I/O.
    pub async fn start(
        &self,
        command: &str,
        opts: ExecStartOpts,
    ) -> Result<(ExecSession, mpsc::UnboundedReceiver<StreamEvent>)> {
        let mut body = serde_json::Map::new();
        body.insert("cmd".into(), Value::String(command.into()));
        if let Some(args) = opts.args {
            body.insert("args".into(), serde_json::to_value(args)?);
        }
        if let Some(env) = opts.env {
            body.insert("envs".into(), serde_json::to_value(env)?);
        }
        if let Some(cwd) = opts.cwd {
            body.insert("cwd".into(), Value::String(cwd));
        }
        if let Some(t) = opts.timeout {
            body.insert("timeout".into(), json!(t));
        }
        if let Some(t) = opts.max_run_after_disconnect {
            body.insert("maxRunAfterDisconnect".into(), json!(t));
        }

        let resp = self
            .ctx
            .http
            .post(format!(
                "{}/sandboxes/{}/exec",
                self.ctx.api_url, self.sandbox_id
            ))
            .headers(self.ctx.headers())
            .json(&Value::Object(body))
            .send()
            .await?;

        #[derive(Deserialize)]
        struct StartResp {
            #[serde(rename = "sessionID")]
            session_id: String,
        }
        let started: StartResp = parse_response(resp, "start exec session").await?;
        self.attach(&started.session_id).await
    }

    /// Alias for [`Exec::start`].
    pub async fn background(
        &self,
        command: &str,
        opts: ExecStartOpts,
    ) -> Result<(ExecSession, mpsc::UnboundedReceiver<StreamEvent>)> {
        self.start(command, opts).await
    }

    pub async fn attach(
        &self,
        session_id: &str,
    ) -> Result<(ExecSession, mpsc::UnboundedReceiver<StreamEvent>)> {
        let ws_base = self
            .ctx
            .api_url
            .replacen("https://", "wss://", 1)
            .replacen("http://", "ws://", 1);
        let auth = if !self.ctx.api_key.is_empty() {
            format!("?api_key={}", urlencoding(&self.ctx.api_key))
        } else {
            String::new()
        };
        let ws_url = format!(
            "{}/sandboxes/{}/exec/{}{}",
            ws_base, self.sandbox_id, session_id, auth
        );

        let (ws_stream, _) = tokio_tungstenite::connect_async(&ws_url).await?;
        let (mut ws_sink, mut ws_read) = ws_stream.split();

        let (events_tx, events_rx) = mpsc::unbounded_channel();
        let (stdin_tx, mut stdin_rx) = mpsc::unbounded_channel::<Vec<u8>>();
        let (exit_tx, exit_rx) = oneshot::channel::<i32>();

        let writer = tokio::spawn(async move {
            while let Some(payload) = stdin_rx.recv().await {
                let mut msg = Vec::with_capacity(1 + payload.len());
                msg.push(0x00);
                msg.extend_from_slice(&payload);
                if ws_sink.send(Message::Binary(msg)).await.is_err() {
                    break;
                }
            }
            let _ = ws_sink.close().await;
        });

        let events_tx_for_reader = events_tx.clone();
        let reader = tokio::spawn(async move {
            let mut got_exit = false;
            let mut exit_tx = Some(exit_tx);
            while let Some(msg) = ws_read.next().await {
                let msg = match msg {
                    Ok(m) => m,
                    Err(_) => break,
                };
                let data = match msg {
                    Message::Binary(b) => b,
                    Message::Close(_) => break,
                    _ => continue,
                };
                if data.is_empty() {
                    continue;
                }
                let stream_id = data[0];
                let payload: Vec<u8> = data[1..].to_vec();
                match stream_id {
                    0x01 => {
                        let _ = events_tx_for_reader.send(StreamEvent::Stdout(payload));
                    }
                    0x02 => {
                        let _ = events_tx_for_reader.send(StreamEvent::Stderr(payload));
                    }
                    0x03 => {
                        let code = if payload.len() >= 4 {
                            i32::from_be_bytes([payload[0], payload[1], payload[2], payload[3]])
                        } else {
                            0
                        };
                        got_exit = true;
                        let _ = events_tx_for_reader.send(StreamEvent::Exit(code));
                        if let Some(tx) = exit_tx.take() {
                            let _ = tx.send(code);
                        }
                    }
                    0x04 => {
                        let _ = events_tx_for_reader.send(StreamEvent::ScrollbackEnd);
                    }
                    _ => {}
                }
            }
            if !got_exit {
                let _ = events_tx_for_reader.send(StreamEvent::Exit(-1));
                if let Some(tx) = exit_tx.take() {
                    let _ = tx.send(-1);
                }
            }
            // Stop the writer once the reader is done (drops stdin_tx eventually).
            drop(events_tx_for_reader);
            let _ = writer.await;
        });

        let session = ExecSession {
            session_id: session_id.to_string(),
            sandbox_id: self.sandbox_id.clone(),
            exit: Mutex::new(Some(exit_rx)),
            cached_exit: Mutex::new(None),
            stdin_tx,
            reader: Mutex::new(Some(reader)),
            ctx: self.ctx.clone(),
        };
        Ok((session, events_rx))
    }

    pub async fn list(&self) -> Result<Vec<ExecSessionInfo>> {
        let resp = self
            .ctx
            .http
            .get(format!(
                "{}/sandboxes/{}/exec",
                self.ctx.api_url, self.sandbox_id
            ))
            .headers(self.ctx.auth_only_headers())
            .send()
            .await?;
        parse_response(resp, "list exec sessions").await
    }

    pub async fn kill(&self, session_id: &str, signal: Option<i32>) -> Result<()> {
        let resp = self
            .ctx
            .http
            .post(format!(
                "{}/sandboxes/{}/exec/{}/kill",
                self.ctx.api_url, self.sandbox_id, session_id
            ))
            .headers(self.ctx.headers())
            .json(&json!({ "signal": signal.unwrap_or(9) }))
            .send()
            .await?;
        check_ok(resp, "kill exec session").await
    }
}

fn urlencoding(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        let c = b as char;
        if c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.' | '~') {
            out.push(c);
        } else {
            out.push_str(&format!("%{:02X}", b));
        }
    }
    out
}

impl Drop for ExecSession {
    fn drop(&mut self) {
        if let Ok(mut guard) = self.reader.try_lock() {
            if let Some(handle) = guard.take() {
                handle.abort();
            }
        }
    }
}
